jni/src/common/Foreign/JNI/Unsafe/Introspection.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Foreign.JNI.Unsafe.Introspection
+ ( getClassName
+ , getSignatures
+ , toText
+ , classGetNameMethod
+ ) where
+
+import Control.Exception (bracket)
+import Control.Monad (forM)
+import Data.Maybe (catMaybes)
+import Data.Singletons
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import qualified Data.Text.Foreign as Text
+import Foreign.JNI.Unsafe.Internal
+import Foreign.JNI.Types
+import qualified Foreign.JNI.String as JNI
+import Prelude hiding (String)
+import System.IO.Unsafe (unsafePerformIO)
+
+-- | The "Class" class.
+kclass :: JClass
+{-# NOINLINE kclass #-}
+kclass = unsafePerformIO $
+  findClass $ referenceTypeName $ sing @('Class "java.lang.Class")
+
+-- | The "Method" class
+kmethod :: JClass
+{-# NOINLINE kmethod #-}
+kmethod = unsafePerformIO $
+  findClass $ referenceTypeName $ sing @('Class "java.lang.reflect.Method")
```

Comment:
kmethod and kclass should be global references (created with `newGlobalRef`), otherwise, they are only valid in the thread that creates them.